### DISCUSSION TITLE
`yield`ing a promise loses type information

### DISCUSSION THREAD

**Discussion by davidatsurge** [mobxjs/mobx-state-tree#1931] [Q&A] [answered]
IIUC, the recommended way to go about async actions in MST is to use generators. But the [approach described in the docs](https://mobx-state-tree.js.org//concepts/async-actions#what-about-async-await) loses type safety, right? Consider:

```ts
import axios from "axios";

type Project = {id: string}

function getProjects() {
  return axios.get<Project[]>("https://.example.com/api/projects")
}

function* generator() {
  const projects = yield await getProjects(); // HERE: the type of `projects` is `any`!
}
```

Is a manual cast required here? Or is there a way to keep type safety?

**Comment by BATCOH** [comment, marked as answer]
[toGenerator](https://mobx-state-tree.js.org/API/#togenerator)
```typescript
import { toGenerator } from "mobx-state-tree";

...

function* generator() {
  const projects = yield* toGenerator(getProjects());
}
```

**Reply by davidatsurge** [reply]
Thank you!
